### ISSUE TITLE
idea: Do not close filemaker between certain steps or workstations

### ISSUE THREAD

**Issue by urapadmin** [arch-kiosk/arch-kiosk-office#309]
or maybe not at all? At least a manual test shows that starting filemaker takes long and can be avoided by having it open already.

**Comment by urapadmin** [comment]
perhaps I could just leave filemaker open for a certain amount of time, say a minute. If no new request for a filemaker object occurs in this time, filemaker will be closed. Needs to be redis controlled because we do not want to close filemaker in one thread because that thread has not received a new request while another thread did receive a request and is even working on it!

**Comment by urapadmin** [comment]
see also #1768

**Comment by urapadmin** [comment]
this is not entirely trivial as FileMaker always shuts down when I close the database. What I need to do here is start a FM Application (with the same privileges that MCP has, so MCP would be the best place to do it) and keep it alive by having it open a dummy fmp12 file. After all the workstations are prepared it can be closed by calling quit on the com application object. So I think this needs MCP to pre-start FileMaker and then eliminate it if no process needs it any longer. I could mark processes that need FileMaker so that MCP knows when to start it and when to stop it. The workstations themselves don't need to start FM at all any longer. I wonder how fickle that'll be.

**Comment by urapadmin** [comment]
Maybe it is possible to teach the export to fm part to wait for FileMaker a little bit if it is not there, yet. So MCP would start FileMaker and then right away start the Kiosk process that needs it. The Kiosk process waits until it is there (or runs fork in the meantime). This solves #1768.

**Comment by urapadmin** [comment]
> Maybe it is possible to teach the export to fm part to wait for FileMaker a little bit if it is not there, yet. So MCP would start FileMaker and then right away start the Kiosk process that needs it. The Kiosk process waits until it is there (or runs fork in the meantime). This solves [#1768](https://github.com/arch-kiosk/arch-kiosk-office/issues/1768).

Yes, but this needs a robust lock using redis. A background thread can start FileMaker while fork etc. is carrying on but once I hit a real FileMaker operation I must wait until an ongoing FM start has finished. So when starting FM I need to set an "initializing" lock before doing anything else. Only when that is gone the  MCP workers are allowed to connect to the FM instance.

**Comment by urapadmin** [comment]
### kiosk 1.7.40
- this solves #1768, too, so that one gets closed as a duplicate. 
- When you prepare docks in bulk FileMaker will be started with the first dock before fork. So it starts in the background.
- FM does not get closed after the first dock (or even the last dock), instead if can be used by the next bulk operation
- FM does get closed (and eradicated, which might solve #3583) before any NON bulk operation
- When the first dock (and all the subsequent docks) reach the "starting filemaker" part at about 50% they should more or less immediately move on to the next step, checking database consistency. That shows, FileMaker was already oipen.
- it can happen that the first dock was so fast during fork, that the filemaker start is still going on. In that case The dock will wait a bit during "starting FileMaker". It actually does not start but wait for FM.
- the subsequent docks should always show "starting FileMaker" only very briefly and move on right away.

This is almost impossible to test in detail. So the test here is "does preparing docks work properly as before"? And that means both bulk and single dock operations. And of course also all other dock operations should work normally. This needs several full test runs of all operations, including sync and we have to keep it under observation for the next months. Even if it does not get stuck or produce obvious errors, please report everything that looks suspicious to you.

Things that can interfere: 
- other users: If you run a bulk preparation of several docks on test, somebody on ANC might sync or do something with their docks. That ANC job will terminate FileMaker (at least as long as ANC itself is not on 1.7.40). The result would be that FileMaker gets closed before that ANC job can run and then the next dock for PVD will have to start it again. That dock would be slower all of a sudden. 

As I said: Impossible to test. So whenever you see things that look sluggish, suspicious etc. not the dock, the project you are on and very important: The exact time. Only then will I be able to look at the several logs and see if things were running concurrently. Since that is a tedious task I will only do that if something is suspicious repeatedly. Keep reporting.

**Comment by urapadmin** [comment]
09.III. 10:10 EST on PVD:
- resetting all docks worked fine for the first and then the second dock is stuck at processing ... which indicates that the lock that protects the FileMaker start (there is none here!) kept the process from continuing. I have seen this before, so the lock management must have a hole.
- after the lock timeout I get 

<img width="896" height="74" alt="Image" src="https://github.com/user-attachments/assets/7d3e30b4-416b-4610-aede-1bfbeb4be0d2" />
 
- the last two docks proceeded just fine
- nothing was going on on ANC at the same time it seemed.

**Comment by luizaogs** [comment]
Preparing all docks worked well and as described above (they basically move on from 50% right away).

**Comment by luizaogs** [comment]
09.III 10:47 PST on PVD:

Synchronization did not work: 

<img width="1479" height="208" alt="Image" src="https://github.com/user-attachments/assets/44a8d283-9f39-4f4a-9d87-2d25e96c0217" />

**Comment by urapadmin** [comment]
> 09.III 10:47 PST on PVD:
> 
> Synchronization did not work:

### MCP 0.6.1 active on Aha
This does not need a new Kiosk version. The active MCP version can be found in any Kiosk under administration. 
The issue here would have affected ANC, too. I try a different strategy (this is all the usual FM prodding, so a lot of try and error). Let's see how that works ...

**Comment by luizaogs** [comment]
Synchronization is working now. So put this under observation?

**Comment by luizaogs** [comment]
(I tested other dock functions on PVD as well, no issues so far)

**Comment by urapadmin** [comment]
@arch-kiosk/test 
Ja, this is rather crazy and experimental. We need to test this under load and over a longer stretch of time. I don't believe in its stability, yet.

**Comment by urapadmin** [comment]
@lbestock It might help to tell active users on the ANC team that we are running something experimental here and that they please report when things don't look or even run normal in the Hub.

**Comment by urapadmin** [comment]
### Kiosk 1.7.40.1 / MCP 0.6.2
- improved the FileMaker handling to address

> 09.III. 10:10 EST on PVD:
> 
>     * resetting all docks worked fine for the first and then the second dock is stuck at processing ... which indicates that the lock that protects the FileMaker start (there is none here!) kept the process from continuing. I have seen this before, so the lock management must have a hole.

**Comment by urapadmin** [comment]
- I can see that when bulk preparing two docks FM gets closed after the first dock. But not after the second. That is weird.

**Comment by luizaogs** [comment]
Hm I don't think that was the case yesterday? They all seemed to be moving on from 50% rather quickly. Unless that's not what you mean.

**Comment by urapadmin** [comment]
> Hm I don't think that was the case yesterday? They all seemed to be moving on from 50% rather quickly. Unless that's not what you mean.

you would not notice. If FM gets closed after the first dock, the second dock will immediately start it before fork again and once that second dock gets to needing FM it would be there. That's really only something I can witness on the server while it is happening .

**Comment by luizaogs** [comment]
🐄

**Comment by lbestock** [comment]
Extremely strange to me: this is what it looks like at the end.

<img width="1520" height="859" alt="Image" src="https://github.com/user-attachments/assets/40ee0c75-d773-48e1-9e0f-1c8b1d502b90" />

**Comment by urapadmin** [comment]
it really helps to get an error message in such a case. That's what the button "see details" is for

**Comment by urapadmin** [comment]
it tells us the cause of this immediately:

<img width="1058" height="59" alt="Image" src="https://github.com/user-attachments/assets/ed5028c1-b952-4542-8f3d-41cc90fda524" />
so not connected to this ticket here, ANC simply needs a new version of the recording app.
 
- new ticket: #3588
